crates/types/src/identifiers.rs

```diff
@@ -238,45 +250,96 @@ pub trait ResourceId {
     }
-}
 
-impl TimestampAwareId for InvocationUuid {
-    fn timestamp(&self) -> MillisSinceEpoch {
-        self.0.timestamp_ms().into()
+    pub fn generate(
+        invocation_target: &InvocationTarget,
+        idempotency_key: Option<&ByteString>,
+    ) -> Self {
+        // --- Rules for deterministic ID
+        // * If the target IS a workflow run, use workflow name + key
+        // * If the target IS an idempotent request, use the idempotency scope + key
+        // * If the target IS NOT an idempotent request, or IS NOT a workflow run, then just generate a random ulid
+
+        let id = match (idempotency_key, invocation_target.invocation_target_ty()) {
+            (_, InvocationTargetType::Workflow(WorkflowHandlerType::Workflow)) => {
+                // Workflow run
+                let mut hasher = Sha256::new();
+                hasher.update(b"wf");
+                hasher.update([Self::HASH_SEPARATOR]);
+                hasher.update(invocation_target.service_name());
+                hasher.update([Self::HASH_SEPARATOR]);
+                hasher.update(
+                    invocation_target
+                        .key()
+                        .expect("Workflow targets MUST contain a key"),
+                );
+                let result = hasher.finalize();
+                let (int_bytes, _) = result.split_at(size_of::<u128>());
+                u128::from_be_bytes(
+                    int_bytes
+                        .try_into()
+                        .expect("Conversion after split can't fail"),
+                )
+            }
+            (Some(idempotency_key), _) => {
+                // Invocations with Idempotency key
+                let mut hasher = Sha256::new();
+                hasher.update(b"ik");
+                hasher.update([Self::HASH_SEPARATOR]);
+                hasher.update(invocation_target.service_name());
+                if let Some(key) = invocation_target.key() {
+                    hasher.update([Self::HASH_SEPARATOR]);
+                    hasher.update(key);
+                }
+                hasher.update([Self::HASH_SEPARATOR]);
+                hasher.update(invocation_target.handler_name());
+                hasher.update([Self::HASH_SEPARATOR]);
+                hasher.update(idempotency_key);
+                let result = hasher.finalize();
+                let (int_bytes, _) = result.split_at(size_of::<u128>());
+                u128::from_be_bytes(
+                    int_bytes
+                        .try_into()
+                        .expect("Conversion after split can't fail"),
+                )
+            }
+            (_, _) => {
+                // Regular invocation
+                Ulid::new().into()
+            }
+        };
+
+        InvocationUuid(id)
     }
 }
 
 impl fmt::Display for InvocationUuid {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let raw: u128 = self.0.into();
+        let raw: u128 = self.0;
```

Comment:
Is this intentional? any difference between the two?

Reply:
it's useless now, before the type was `Ulid`, that's why the `into()`